### ISSUE TITLE
Everything choppy after `xset dpms force off`

### ISSUE THREAD

**Issue by tsujan** [compiz-reloaded/compiz#95]
I have this problem for several months. It doesn't happen with kwin, Compton or Enlightenment's compositor: only with Compiz.

I may turn off the monitor with `xset dpms force off`, through a shortcut, or the monitor may be turned off after 7 min of inactivity because I have `xset dpms 0 0 420` in my startup. However, when the monitor is turned on by a mouse movement or a key press, *sometimes* everything becomes choppy in Compiz. Jerky animations go away only by restarting Compiz (I've even assigned a shortcut to that).

This issue happens quite randomly -- I couldn't find any pattern about it -- but the effect is always the same: extremely jerky animations.

**Comment by soreau** [comment]
@tsujan Maybe compiz is processing data unnecessarily. Does the compiz process also consume more cpu cycles than normal, when it gets into this choppy animation state?

**Comment by tsujan** [comment]
> Does the compiz process also consume more cpu cycles than normal, when it gets into this choppy animation state?

Since I'm sensitive to high CPU usage and have 2 kinds of CPU monitors, I don't think the CPU time is taken on such occasions. On the other hand, this is a powerful laptop with Intel Core i7.

However, to be certain about that, I'll report back when it happens again.

**Comment by tsujan** [comment]
It happened again and there was no extra CPU usage (%1) but this time, I found something interesting: Instead of restarting Compiz, I opened CCSM and unchecked "Sync to VBlank" under "Display Settings". After that, the choppy animations went away completely :)

I can't see any side-effect (no tearing) but it's still too soon to be sure. "Sync to VBlank" was always checked on an Asus laptop with similar specs and when I started to use this Lenovo Ideapad (because Asus's LCD cable had a bad problem), I just moved the Compiz settings to it.

Why synching to vblank works before `xset dpms force off` but, sometimes, not after it, I don't know. The integrated graphic card is Intel, which I always use. I have bumblebee and had uninstalled `xf86-video-intel` because, unlike in Asus, it caused serious problems, which disappeared altogether after its uninstallation.

**Comment by tsujan** [comment]
I always thought I had smooth animations with Compiz Reloaded. Now, *without* vsync, I see what real smoothness is -- like old days of Beryl.

Sorry for the noise. Closing this.

**Comment by tsujan** [comment]
Reopening this because disabling "Sync to VBlank" caused video tearing. For some reason, it doesn't work OK after `xset dpms force off`.

**Comment by tsujan** [comment]
I tested a screensaver (xscreensaver) and also screen blanking instead of turning off the screen and the vblank sync was always OK. Apparently, its problem appears only when the screen is turned off.

**Comment by soreau** [comment]
@tsujan Maybe it's a driver bug. You mentioned xf86-video-intel was causing problems and you're using bumblebee.. have you tried xf86-video-modesetting?

**Comment by tsujan** [comment]
> You mentioned xf86-video-intel was causing problems and you're using bumblebee.. have you tried xf86-video-modesetting?

I use xorg-server's modesetting (under Manjaro and as recommended by Arch).

Here, Compiz is as smooth as it was on the other laptop (Asus). This doesn't happen without vblank synching but once vblank synching is disabled, tearing is visible in video playing and also in the wobbly effect.

I took a look at code. `waitForVideoSync()` (in `display.c`) seemed OK. I wonder how Compiz could detect that the screen off -- couldn't find anything about that.

**Comment by soreau** [comment]
It looks like the path where a screen change happens would start around event.c:1313 where it calls configureScreen() which ultimately calls updateOutputDevices(), both defined in screen.c. I am doubting this is a problem with compiz but more likely a driver bug compiz exposes. There are a couple options in workarounds plugin you might check out though, the ones that mention XGL and GLX.

**Comment by tsujan** [comment]
> It looks like the path where a screen change happens would start around...

Thanks for the explanation!

> more likely a driver bug compiz exposes

That's possible.

> There are a couple options in workarounds plugin you might check out

I'll try them.

**Comment by tsujan** [comment]
Forcing sync between X and GLX seems to work; I can't reproduce the issue after enabling workarounds only with it. Unlike what its tooltip says -- and that was why I hadn't tried it before -- it can't be just for Nvidia.

So, this may be a driver problem, as you mentioned.

Thank you for your attention to this issue!

**Comment by tsujan** [comment]
Ah, this is so random! It happened again, after 20 min of `xset dpms force off` and *with* the above-mentioned workaround. The other XGL/GLX workarounds had no effect either. The only thing that had effect was the removal of vblank sync -- which can't be used because of tearing. When I enabled it again, choppiness came back. Restarting Compiz was needed.

BTW, using Kernel 4.14 (LTS and latest stable) instead of the previous LTS (4.9.63) had no tangible effect either.

**Comment by tsujan** [comment]
As for the driver problem, I reverted to the Intel driver from modesetting (`xf86-video-intel` doesn't seem so buggy anymore) but the problem persists.

**Comment by tsujan** [comment]
Replacing DRI3 with DRI2 in modesetting helped. No issue since then.

There's an interesting difference between DRI3 and DRI2 in this regard. When the monitor is going to be turned on, it happens suddenly with DRI3 but there's a short and nice fade-in with DRI2.

**Comment by raveit65** [comment]
Btw. nvidia driver doesn't support DRI3 garbage.

**Comment by tsujan** [comment]
> Btw. nvidia driver doesn't support DRI3 garbage.

As I said above, my integrated card is Intel.

**Comment by tsujan** [comment]
Conclusion: Some problems aren't in Compiz but revealed by it, as @soreau mentioned. Modesetting or Intel, DRI3 has problems.

This can be closed, I think.

**Comment by tsujan** [comment]
In case someone with a similar problem comes to this page:

With mosesetting/Intel + DRI3, glxgears might show 70-90 FPS in the first 5 sec, although it became ~60 afterwards. With DRI2, it's always 60 FPS.

**Comment by tsujan** [comment]
More than a week has passed and Compiz has worked OK with DRI2. However, (1) Full-screen videos should be unredirected; otherwise strange things that didn't exist with DRI3 will happen; (2) if I want to log into Enlightenment, I should switch to DRI3; E's compositor works smoothly only with DRI3.

Since the new Compiz's code is rooted in the old Compiz's, this question comes to my mind: Is it possible that its code is "tuned" to DRI2 and not ready for DRI3, in some sense?

**Comment by soreau** [comment]
Even though compiz-reloaded is quite DRI1 era, mesa should do The Right Thing depending on which version of DRI is loaded.

**Comment by tsujan** [comment]
@soreau Thanks for the informative answer!

**Comment by soreau** [comment]
For what it's worth, when I tried `xset dpms force off`it didn't cause choppy animations but it did cause default settings to be used for cube transparency while rotating and zoom, but only when screensaver plugin initiated the cube rotate. I had to restart X before it began behaving correctly again (even after restarting compiz and running `xset dpms force on` it was broken). Not sure what this does but it seems to mess things up in strange ways.

**Comment by tsujan** [comment]
Interesting! With DRI3?

Since I selected DRI2 here, I enjoy Compiz and `xset dpms force off`. The Intel driver causes very smooth animations -- modesetting is good too but a little choppy sometimes. I also enabled Compiz's screensaver once to see its CPU usage but no issue happened.

That's while DRI2 is terrible with Enlightenment but DRI3 is fine there.

However, as far as I've seen, even slightly different cards may show very different behaviors.

**Comment by tsujan** [comment]
There's definitely a bug somewhere -- probably, in mesa. If only we could report it....

**Comment by soreau** [comment]
@tsujan Yes, with DRI3 on amdgpu+radeonsi. I did not test DRI2.

**Comment by JockeTF** [comment]
I can confirm this bug as well.

```
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) Ivybridge Mobile  (0x166)
    Version: 17.2.6
    Accelerated: yes
    Video memory: 1536MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.2
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Ivybridge Mobile 
OpenGL core profile version string: 4.2 (Core Profile) Mesa 17.2.6
OpenGL core profile shading language version string: 4.20
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.0 Mesa 17.2.6
OpenGL shading language version string: 1.30
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 17.2.6
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
```

**Comment by tsujan** [comment]
I think mesa has hidden bugs that aren't reported because they can't be detected without making comparisons. Another example: https://bugs.kde.org/show_bug.cgi?id=382817; it turned out that the problem was (mostly) in dri3 but I couldn't know that if I didn't have Compiz. Also, people may be happy with Compiz+dri3 just because they think that animations can't be smoother -- as is the case with kwin...

**Comment by tsujan** [comment]
The main problem discussed here seems to be fixed with mesa-17.3.0 and modesetting+dri3.
